### DISCUSSION TITLE
Implementing a Source and Sink with I2s

### DISCUSSION THREAD

**Discussion by qcorr** [pschatzmann/ESP32-A2DP#409] [Q&A] [answered]
Hello,
I want to first start off and say I greatly appreciate your library.
I am trying to use it to have one esp32 as a source and communicate with another esp32 as a sink.
I want to read BT audio data in from a phone on one chip and write the data to gpio pins using I2s.
The other chip should read in the data from its gpio pins and write to a bluetooth speaker. I have already tested your example code and the chip connects to the speaker and sends the sin wave audio.
I think my issue is with the get_data_channels method and how I am filling the audio buffer. If you have any idea how to correct this, please let me know. Currently, the audio being sent from the chip to the speaker is choppy and unrecognizable.
Attached is the code used.

first chip - receive data from phone and send to GPIO pins using DAC
//libraries
#include "BluetoothA2DPSink.h"

//create sink object
BluetoothA2DPSink a2dp_sink;

void setup() {
    //override i2s config
    //sets output to DAC pins GPIO25 (channel 1) and GPIO26 (channel 2)
    static const int i2s_num = 0; // i2s port number

    static const i2s_config_t i2s_config = {
        .mode = (i2s_mode_t) (I2S_MODE_MASTER | I2S_MODE_TX | I2S_MODE_DAC_BUILT_IN),
        .sample_rate = 44100, // corrected by info from bluetooth
        .bits_per_sample = (i2s_bits_per_sample_t) 16, /* the DAC module will only take the 8bits from MSB */
        .channel_format = I2S_CHANNEL_FMT_RIGHT_LEFT,
        .communication_format = (i2s_comm_format_t)I2S_COMM_FORMAT_STAND_MSB,
        .intr_alloc_flags = 0, // default interrupt priority
        .dma_buf_count = 8,
        .dma_buf_len = 64,
        .use_apll = true
    };

    i2s_driver_install(I2S_NUM_0, &i2s_config, 0, NULL);

    //create AUXSYNQ instance
    a2dp_sink.set_i2s_config(i2s_config);
    a2dp_sink.start("AUXSYNQ");
    

}

void loop() {
}

Second chip - read from gpio and send to speaker
//libraries
#include "BluetoothA2DPSource.h"

//pins to receive input from
#define AUDIO_PIN1 34
#define AUDIO_PIN2 35

//create source object to send to speakers
BluetoothA2DPSource a2dp_source;

double amplitude = 1000.0;

// The supported audio codec in ESP32 A2DP is SBC. SBC audio stream is encoded
// from PCM data normally formatted as 44.1kHz sampling rate, two-channel 16-bit sample data
int32_t get_data_channels(Frame *frame, int32_t channel_len) {
    //generate audio source
    uint16_t buffer[channel_len];
    for(int i = 0; i < channel_len; i++)
    {
      buffer[i] = (analogRead(AUDIO_PIN1) + analogRead(AUDIO_PIN2)) / 2;
      //buffer[i] = (adc1_get_raw(ADC1_CHANNEL_6) + adc1_get_raw(ADC1_CHANNEL_7)) / 2; //GPIO 34 and 35
    }
    // fill the channel data IF we have data
    for (int sample = 0; sample < channel_len; ++sample) {
        frame[sample].channel1 = buffer[sample] * amplitude;
        frame[sample].channel2 = frame[sample].channel1;
    }

    return channel_len;
}

void setup() {
    //connect ESP32 to speaker
    analogReference(DEFAULT);
    a2dp_source.start("SRS-XB33", get_data_channels); 
    a2dp_source.set_volume(100);
}

void loop() {
}

**Comment by pschatzmann** [comment, marked as answer]
This is definitely not an issue, so I converted it to a discussion. Please use issues for bugs only! 

Then, from your description you left me totally confused and I don't quite understand what you try to do. 

Looking at your code, there is no way your approach can work. Here are the issues
- You need to sample audio with a consistent rate (in this case with 44100 samples per second), so analog read is completely wrong here. I suggest that you have a look at the [AudioTools](https://github.com/pschatzmann/arduino-audio-tools) project and you can try to use the AnalogAudioStream. But before you can even think of reading and transmitting any audio data you need to build a test case to make sure that you have a correct signal.
-  If I understood you correctly, you try to transmit audio over wires: I am not sure what the best approach here is: using analog, the quality will be just horrible and you will get a lot of noise!  
- I was experimenting with sending audio via Serial but w/o success. 
- Others have tried using I2s, but they were running into issues as well.

And finally here are my [guidelines if something is not working](https://github.com/pschatzmann/arduino-audio-tools/wiki/Encoding-and-Decoding-of-Audio).

**Reply by qcorr** [reply]
Thank you for the reply and for switching the topic to discussion instead of issue, my apologies.
Yes, I am trying to send audio over wires between gpio pins on two seperate ESP32s.
The first chip needs to take the audio from the phone and output it to external DAC. Using the code provided for the library, it already achieves this and sends it to pins 25 and 26.
After that, I have wires from pins 25 and 26 to the second esp32 where I would like to read in the audio data and send it to a speaker. I hope this better describes what I am trying to do.

"If I understood you correctly, you try to transmit audio over wires: I am not sure what the best approach here is: using analog, the quality will be just horrible and you will get a lot of noise!"
Any ideas on another approach then for what I want to do? Maybe an amplifier module, etc?

If I do try and use AnalogAudioStream, how would I modify the get_data_channels to fill the audio buffer with that instead? Is the approach above valid or is there another best practice to sending the audio into the frames?

Thank you for your reply, my friends and I are college students working on a personal project!

**Comment by pschatzmann** [comment]
- I am not sure how to make stereo work, so you might need to modify my code. If you find a solution please share it. But I suggest that you try out mono first, to find out what the quality will be.
- I never found any good solution to your problem so you will have quite a big challenge:
  - you could try i2s, but people are reporting that this has issues
  - Serial seems to be slow and unreliable

**Reply by dotanger** [reply]
Hi,
I'm also interested in a similar implementation. I'm new to the audio aspects.
I wanted one esp as sink from phone and the other as source to speaker.
Then the question is how to transfer the audio data from the sink esp to the source esp via wires.
You write people have tried using i2s and report issues.
What about SPI?
What are the issues reported with i2s?

Thanks a lot

**Reply by pschatzmann** [reply]
I suggest that you have a look at the AudioTools communications examples and try them out yourself.
Please also have a [look at this](https://github.com/pschatzmann/arduino-audio-tools/discussions/149)
